### PR TITLE
mgr/dashboard: display placement column in service table

### PR DESCRIPTION
**Pull request by votdev**
Bildschirmfoto vom 2020-11-23 10-48-31

Fixes: https://tracker.ceph.com/issues/44404

Signed-off-by: Volker Theile <vtheile@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/placement.pipe.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      return $localize`no spec`;
+    }
+    if (_.get(serviceSpec, 'unmanaged', false)) {
+      return $localize`unmanaged`;
+    }
+    const kv: Array<any> = [];
+    const hosts: Array<string> = _.get(serviceSpec, 'placement.hosts');
+    const count: number = _.get(serviceSpec, 'placement.count');
+    const label: string = _.get(serviceSpec, 'placement.label');
+    const hostPattern: Array<string> = _.get(serviceSpec, 'placement.host_pattern');
```
Comment (bk201) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/placement.pipe.ts:26]: The `host_pattern` is a string.

e.g.

Row:
Screen Shot 2020-11-23 at 4 43 53 PM

Spec:
Screen Shot 2020-11-23 at 4 43 42 PM

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/placement.pipe.spec.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        }
+      })
+    ).toBe('label:foo');
+  });
+
+  it('transforms placement (5)', () => {
+    expect(
+      pipe.transform({
+        placement: {
+          host_pattern: ['foo', 'bar']
```
Comment (bk201) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/placement.pipe.spec.ts:62]: `host_pattern ` should accept a string.

**Review by bk201 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/placement.pipe.ts (hunk 1 above)
Comment (votdev) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/placement.pipe.ts:26]: Hmmm, need to investigate that what arrives in the UI. In the [orchestrator](https://github.com/ceph/ceph/blob/2f0fe914be27463d7a5a6b392fcb3adc74a50d72/src/python-common/ceph/deployment/service_spec.py#L222) the host is an array.

**Review by nizamial09 — APPROVED** [review]
Tested Locally! LGTM

**Comment by s0nea** [comment]
jenkins test make check

**Review by bk201 — APPROVED** [review]
Nice feature. Moving placement column forward also looks better.

**Comment by epuertat** [comment]
@votdev minor suggestion: what about this pipe formatting those as badges with a different color for each of the 4 kinds? I think Dashboard tables are _too_ textual, and any visual hint will help make them more visually appealing.

**Review by avanthakkar — APPROVED** [review]
LGTM!

**Review by s0nea — APPROVED** [review]
LGTM!

**Comment by votdev** [comment]
> @votdev minor suggestion: what about this pipe formatting those as badges with a different color for each of the 4 kinds? I think Dashboard tables are _too_ textual, and any visual hint will help make them more visually appealing.

I discussed this with @bk201 already and we came to the conclusion that we should keep the current implementation because using badges will consume too much space.

STrkt3I

We discussed using spaces here to allow the browser to wrap longer lines, but this looks ugly, too. We also came to the conclusion to keep it as it is.

MiKFu7V

**Pull request merged**